iib/web/api_v1.py

```diff
@@ -460,3 +460,108 @@ def regenerate_bundle_batch():
         ', '.join(request_id_strs),
     )
     return flask.jsonify(request_jsons), 201
+
+
+@api_v1.route('/builds/add-rm-batch', methods=['POST'])
+@login_required
+def add_rm_batch():
+    """
+    Submit a batch of requests to add or remove operators from an index image.
+
+    :rtype: flask.Response
+    :raise ValidationError: if required parameters are not supplied
+    """
+    payload = flask.request.get_json()
+    Batch.validate_batch_request_params(payload)
+
+    batch = Batch(annotations=payload.get('annotations'))
+    db.session.add(batch)
+
+    requests = []
+    # Iterate through all the build requests and verify that the requests are valid before
+    # committing them and scheduling the tasks
+    for build_request in payload['build_requests']:
+        try:
+            if build_request.get('operators'):
+                # Check for the validity of a RM request
+                request = RequestRm.from_json(build_request, batch)
+            elif build_request.get('bundles'):
+                # Check for the validity of an Add request
+                request = RequestAdd.from_json(build_request, batch)
+            else:
+                raise ValidationError('Build request is not a valid Add/Rm request.')
+        except ValidationError as e:
+            # Rollback the transaction if any of the build requests are invalid
+            db.session.rollback()
+            raise ValidationError(
+                f'{str(e).rstrip(".")}. This occurred on the build request in '
+                f'index {payload["build_requests"].index(build_request)}.'
+            )
+        db.session.add(request)
+        requests.append(request)
+
+    db.session.commit()
+    messaging.send_messages_for_new_batch_of_requests(requests)
```

Comment:
What happens if we're interrupted after the DB commit and before the messages are sent (or if message sending fails) - does the system ever detect and recover from the "lost" requests?

Reply:
So if an error is encountered while sending messages, it is considered as a non-fatal error and the request continues its course. The delivery of messages is not guaranteed by IIB.